cxx/distributions/normal.hh

```diff
@@ -37,19 +59,45 @@ public:
         var -= (x - mean) * (x - old_mean);
     }
 
+    double sprime() const {
+      // r' = r + N
+      // m' = (r m + N mean) / (r + N)
+      // C = N (var + mean^2)
+      // s' = s + C + r m^2 - r' m' m'
+      double mdelta = r * (m - mean) / (r + N);
+      double mprime = mean + mdelta;
+      return s + r * (m * m - mprime * mprime)
+          + N * (var - 2 * mean * mdelta - mdelta * mdelta);
+    }
+
     double logp(double x) const {
-        double y = (x - mean);
-        return -0.5 * (y * y / var + log(var) + log(M_2PI));
+      return -0.5 * N * log(M_2PI)
+          + logZ(r + N, v + N, sprime())
+          - logZ(r, v, s);
     }
```

Comment:
I haven't checked the math carefully yet, but should this depend in some way on `x`? (Is it possible that what you're computing here is actually the logp_score?)

Here is a reference implementation from Feras: https://github.com/probcomp/cgpm/blob/56a481829448bddc9cdfebd42f65023287d5b7c7/src/primitives/normal.py#L166-L173

Reply:
Fixed.  Both Feras's code and my code are based on the write up at http://www.stats.ox.ac.uk/~teh/research/notes/GaussianInverseGamma.pdf , and I didn't notice that the numerator Z function had a hidden dependence on x.

(I also think Feras's code is missing a factor of N from the log(2 pi) term.)

Reply:
Fixed.  (I had been implementing equation (11) on GaussianInverseGamma.pdf, when I should have been implementing equation (13)).